clippy_lints/src/dereference.rs

```diff
@@ -920,6 +931,115 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
     v.0
 }
 
+// Checks whether:
+// * child is an expression of the form `&e` in an argument position requiring an `impl Trait`
+// * `e`'s type implements `Trait` and is copyable
+// If the conditions are met, returns `Some(Position::ImplArg(..))`; otherwise, returns `None`.
+//   The "is copyable" condition is to avoid the case where removing the `&` means `e` would have to
+// be moved, but it cannot be.
+fn needless_borrow_impl_arg_position<'tcx>(
+    cx: &LateContext<'tcx>,
+    parent: &Expr<'tcx>,
+    child_arg_index: usize,
+    child_arg_ty: Ty<'tcx>,
```

Comment:
Is there an easy way to obtain a `ParamTy` here?

The `ParamTy` would replace both `child_arg_index` and `child_arg_ty` arguments, correct? If so, how would you imagine this line changing? https://github.com/rust-lang/rust-clippy/blob/2507d2ad883b56acd28506f34f4ae6918a32173c/clippy_lints/src/dereference.rs#L996

Reply:
It would only replace `child_arg_ty`. Would look like
```rust
if let ty::Param(ty) = ty {
    needless_borrow_impl_arg_position(...)
} else {
    param_auto_deref_stability(...)
}
```
You would have to change `needless_borrow_impl_arg_position` to return `Position::Other` which would be correct for any generic param.

Reply:
Are there no cases where `ty` is a `Param` and `needless_borrow_impl_arg_position` would return `None` but `param_auto_deref_stability` returns `Some(..)`? Put another way, are we sure it is okay to elide the current code's `unwrap_or_else`?

Separately, could you please explain why this change necessitates the use of `Position::Other`? I am currently using `Position::ImplArg` to determine the error message: https://github.com/rust-lang/rust-clippy/blob/2507d2ad883b56acd28506f34f4ae6918a32173c/clippy_lints/src/dereference.rs#L340-L341

Reply:
`param_auto_deref_stability` returns `Position::Other(_)` for generic types.

There's no need to return an `Option` from `needless_borrow_impl_arg_position `. Either you can remove borrows, or nothing interesting can be done with the location the expression is in. The latter case uses `Position::Other(_)`.